Update hack/update-generated-docs.sh to generate man pages

juanvallejo opened the pull request:
Fixes #9348 
man pages are generated in `docs/man/oc`, `docs/man/openshift`, `docs/man/oadm`

mkumatag commented on tools/genman/gen_oadm_man.go:
```diff
@@ -0,0 +1,54 @@
+
+func GenOadmMan(cmdName string) {
+	// use os.Args instead of "flags" because "flags" will mess up the man pages!
+	path := "docs/man/" + cmdName
```
All 3 functions(GenOadmMan, GenOcMan, GenOpenshiftMan) I see exact same code, can we just write one common function to generate the markdown and call with commandname to it.?

mkumatag commented on tools/genman/gen_man.go:
```diff
@@ -0,0 +1,154 @@
+	}
+}
+
+func GenMarkdown(command *cobra.Command, parent, docsDir string) {
```
All the functions GenMarkdown, GenOadmMan, GenOcMan, GenOpenshiftMan : no need to export them as we are not using them outside, so can be converted into small letter.

juanvallejo commented:
@mkumatag 
> All 3 functions(GenOadmMan, GenOcMan, GenOpenshiftMan) I see exact same code, can we just write one common function to generate the markdown and call with commandname to it.?

Thanks for the feedback, done: https://github.com/openshift/origin/pull/9448/commits/1cee2117640a6f17a71874348d3a5eac8fd10fc1

juanvallejo commented:
[test]

fabianofranz commented:
@sdodson something we might also include in the rpm packaging.
@brenton I remember you were interested in this.

fabianofranz commented:
@juanvallejo We might also update `hack/verify-generated-docs.sh`. That's what will check and make sure tests fail if people propose a PR with docs outdated. Apart that looking great, squash the last 3 commits and let's get this in.

juanvallejo commented:
@fabianofranz 
> We might also update hack/verify-generated-docs.sh. That's what will check and make sure tests fail if people propose a PR with docs outdated

I have updated `hack/verify-generated-docs.sh` to do this, however some commands implement unique or live data that changes every time new man pages are generated, for example https://github.com/juanvallejo/origin/blob/1cee2117640a6f17a71874348d3a5eac8fd10fc1/docs/man/oadm/oadm-create-master-certs.1#L108 this line on the linked man page is generated here: https://github.com/openshift/origin/blob/master/pkg/cmd/server/admin/create_mastercerts.go#L107

Since the contents of the function `DefaultSignerName` change every time the function is called, the `hack/verify-generated-docs.sh` script will fail every time.

The following is a list of all of the files that dynamically change every time they are generated:

```
docs/man/oadm/oadm-ca-create-master-certs.1
docs/man/oadm/oadm-ca-create-signer-cert.1
docs/man/oadm/oadm-create-master-certs.1
docs/man/oadm/oadm-create-signer-cert.1
docs/man/oc/oc-adm-ca-create-master-certs.1
docs/man/oc/oc-adm-ca-create-signer-cert.1
docs/man/oc/oc-adm-create-master-certs.1
docs/man/oc/oc-adm-create-signer-cert.1
docs/man/openshift/openshift-admin-ca-create-master-certs.1
docs/man/openshift/openshift-admin-ca-create-signer-cert.1
docs/man/openshift/openshift-admin-create-master-certs.1
docs/man/openshift/openshift-admin-create-signer-cert.1
docs/man/openshift/openshift-cli-adm-ca-create-master-certs.1
docs/man/openshift/openshift-cli-adm-ca-create-signer-cert.1
docs/man/openshift/openshift-cli-adm-create-master-certs.1
docs/man/openshift/openshift-cli-adm-create-signer-cert.1
```

juanvallejo commented:
@fabianofranz I have updated the `hack/update-generated-docs.sh` script to also update man page when it is run. Due to the issue stated in my last comment, do you think we could simply rely on `hack/verify-generated-docs.sh` checking for up-to-date docs, and keeping man pages up to date whenever it detects that normal docs are out of date?

smarterclayton commented:
We need to fix the commands so that their help output doesn't change from
time to time.  You can probably do that by overriding specific flags from
the command tree prior to generation.

On Tue, Jun 21, 2016 at 6:27 PM, Juan Vallejo <notifications@github.com>
wrote:

> @fabianofranz <https://github.com/fabianofranz> I have updated the
> hack/update-generated-docs.sh script to also update man page when it is
> run. Due to the issue stated in my last comment, do you think we could
> simply rely on hack/verify-generated-docs.sh checking for up-to-date
> docs, and keeping man pages up to date whenever it detects that normal docs
> are out of date?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/9448#issuecomment-227591297>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABG_p2myzFj0k4s7RssqoJD2At4KJ99Nks5qOGVMgaJpZM4I6J0X>
> .
>

juanvallejo commented:
re[test]

juanvallejo commented:
@fabianofranz @smarterclayton Done! I I use these lines to target dynamic flag default values: https://github.com/juanvallejo/origin/blob/jvallejo_gen-man-pages/tools/genman/gen_man.go#L103-107

smarterclayton commented:
That should really be abstracted out into its own function that you inject
down - i.e. none of the deep tools code should know anything about specific
flags, instead the very top level says "here's a function that mutates the
flags to make them idempotent" and the lower levels blindly invoke that
function.

On Wed, Jun 22, 2016 at 11:25 AM, Juan Vallejo <notifications@github.com>
wrote:

> @fabianofranz <https://github.com/fabianofranz> @smarterclayton
> <https://github.com/smarterclayton> Done! I I use these lines to target
> dynamic flag default values:
> https://github.com/juanvallejo/origin/blob/jvallejo_gen-man-pages/tools/genman/gen_man.go#L103-107
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/9448#issuecomment-227780044>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABG_p-Pyu4003230aGC0oTgW9eZsViqkks5qOVPkgaJpZM4I6J0X>
> .
>

juanvallejo commented:
@smarterclayton Thanks for the feedback on this, the `printFlags` function on `tools/genman/gen_man.go` now checks to see if flags have annotations with a key of `manpage-def-value`. If this annotation is present, the default value for that flag is overwritten with the annotation key's value on the manpage.

smarterclayton commented on Godeps/_workspace/src/k8s.io/kubernetes/cmd/genutils/genutils.go:
```diff
@@ -0,0 +1,41 @@
+limitations under the License.
+*/
+
+package genutils
```
Why does this need to be upstream?

juanvallejo replied:
@smarterclayton It's present in newer versions of k8s: https://github.com/kubernetes/kubernetes/tree/master/cmd/genutils

smarterclayton replied:
OK.

On Wed, Jun 22, 2016 at 2:16 PM, Juan Vallejo <notifications@github.com>
wrote:

> In Godeps/_workspace/src/k8s.io/kubernetes/cmd/genutils/genutils.go
> <https://github.com/openshift/origin/pull/9448#discussion_r68106188>:
>
> > +Copyright 2015 The Kubernetes Authors All rights reserved.
> > +
> > +Licensed under the Apache License, Version 2.0 (the "License");
> > +you may not use this file except in compliance with the License.
> > +You may obtain a copy of the License at
> > +
> > +    http://www.apache.org/licenses/LICENSE-2.0
> > +
> > +Unless required by applicable law or agreed to in writing, software
> > +distributed under the License is distributed on an "AS IS" BASIS,
> > +WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> > +See the License for the specific language governing permissions and
> > +limitations under the License.
> > +*/
> > +
> > +package genutils
>
> @smarterclayton <https://github.com/smarterclayton> It's present in newer
> versions of k8s:
> https://github.com/kubernetes/kubernetes/tree/master/cmd/genutils
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/9448/files/1c1c7e7d56c397f95ad0f520bee03751ee160c72#r68106188>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABG_p1r2WHZWvEY7DJLYUhl46nQLStLIks5qOXwLgaJpZM4I6J0X>
> .
>

smarterclayton commented:
Please split your generator code commit from the generated code commit.

juanvallejo commented:
@fabianofranz @smarterclayton 
> Please split your generator code commit from the generated code commit.
Done!

juanvallejo commented:
[test]

fabianofranz commented:
LGTM, @smarterclayton do you have anything else? @sdodson needs RPM packaging follow-up.

smarterclayton commented on hack/verify-generated-docs.sh:
```diff
@@ -16,6 +16,10 @@ GENERATED_DOCS_ROOT="${OS_ROOT}/${GENERATED_DOCS_ROOT_REL}"
 TMP_GENERATED_DOCS_ROOT_REL="_output/verify-generated-docs"
```
These tasks need to be added to `make verify` and `make update` or they won't be run.

juanvallejo replied:
@smarterclayton 
>These tasks need to be added to make verify and make update or they won't be run.

The `hack/verify-generated-docs.sh` task was are already there from previous contributions, I simply updated it so that it verifies man pages as well

sdodson commented:
> @sdodson needs RPM packaging follow-up.

we'll just roll it into the same issue where we're moving bash completions into rpmbuild process

smarterclayton commented:
Ok, lgtm.

On Thu, Jun 23, 2016 at 4:38 PM, Scott Dodson <notifications@github.com>
wrote:

> @sdodson <https://github.com/sdodson> needs RPM packaging follow-up.
>
> we'll just roll it into the same issue where we're moving bash completions
> into rpmbuild process
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/9448#issuecomment-228176340>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABG_p61-QgzCPhQtxbWpxbg9zmODKcNUks5qOu7TgaJpZM4I6J0X>
> .
>

fabianofranz commented:
[merge]

fabianofranz commented:
@juanvallejo `FAILURE: Generated docs up to date, but generated man pages out of date. Please run hack/update-generated-docs.sh`

juanvallejo commented:
@fabianofranz 
>FAILURE: Generated docs up to date, but generated man pages out of date. Please run hack/update-generated-docs.sh

Thanks! Docs and manpages should verify with no issues now

smarterclayton commented:
```

+++ /data/src/github.com/openshift/origin/_output/verify-generated-docs/docs/man/openshift/openshift-cli-expose.1	2016-06-25
06:58:43.000000000 -0400
@@ -86,7 +86,7 @@ labels from the object it exposes.

 .PP
 \fB\-\-port\fP=""
-    The port that the service should serve on. Copied from the
resource being exposed, if unspecified
+    The port that the resource should serve on.

 .PP
 \fB\-\-protocol\fP=""
FAILURE: Generated docs up to date, but generated man pages out of
date. Please run hack/update-generated-docs.sh

```

On Sat, Jun 25, 2016 at 7:50 AM, OpenShift Bot <notifications@github.com>
wrote:

> continuous-integration/openshift-jenkins/merge FAILURE (
> https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/5419/)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/9448#issuecomment-228535769>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABG_p765r9mCtil2JvL2wuBVsQDUTcm4ks5qPRX7gaJpZM4I6J0X>
> .
>

juanvallejo commented:
@smarterclayton Thanks, it seems the help description of the `oc expose` command was updated before this PR was tested. I ran `./hack/verify-genereated-docs.sh` locally, receiving no errors. Once I updated my master branch and did a rebase, I received this same diff output locally. I have tracked the new changes, squashed, and updated this PR

smarterclayton commented:
[merge]

smarterclayton commented:
This is adding some flags to man pages that come from the base Golang flags package, but are not present in our actual commands:

```
-\fB\-\-storage\-driver\-buffer\-duration\fP=0
-    Writes in the storage driver will be buffered for this duration, and committed to the non memory backends as a single transaction
-
-.PP
-\fB\-\-storage\-driver\-db\fP="cadvisor"
-    database name
-
-.PP
-\fB\-\-storage\-driver\-host\fP="localhost:8086"
-    database host:port
-
-.PP
-\fB\-\-storage\-driver\-password\fP="root"
-    database password
-
-.PP
-\fB\-\-storage\-driver\-secure\fP=false
-    use secure connection with database
-
-.PP
-\fB\-\-storage\-driver\-table\fP="stats"
-    table name
-
-.PP
-\fB\-\-storage\-driver\-user\fP="root"
-    database username
-
-.PP
```

If you compile on a Mac or Windows these flags are excluded (because they are part of packages that are only available on Mac).  

Please open an issue and make sure these are excluded - @fabianofranz can help you identify the ones to remove.

eparis commented:
If this generated them on the docs it means the flags are real flags that would be parsed and would go $somewhere...

smarterclayton commented:
The rule of thumb is that we don't use global flags.  All the ones
here have replacements or should be explicitly copied in (like we do
for glog).
